### ISSUE TITLE
ARM sigbus from initialiseAOTRelocationHeader

### ISSUE THREAD

**Issue by JamesKingdon** [eclipse/openj9#1205]
J9::ARM::AheadOfTimeCompile::initialiseAOTRelocationHeader(TR::IteratedExternalRelocation*) uses stm to access unaligned memory causing sigbus and a core file when running on 64bit OS. This causes a failure in test/VM_Test

`=> 0xf5d8ee78 <_ZN2J93ARM18AheadOfTimeCompile29initialiseAOTRelocationHeaderEPN2TR26IteratedExternalRelocationE+1504>:	stmib	r4, {r1, r3}`

```
(gdb) info reg
r0             0xf0f0f6e8	4042323688
r1             0xffffffff	4294967295
r2             0xed13f102	3977507074
r3             0xee008df0	3993013744
r4             0xed13f0f6	3977507062
r5             0xf0c05608	4039136776
r6             0xf6399c78	4130970744
r7             0xefc00000	4022337536
r8             0xefc076c0	4022367936
r9             0xf11abcd8	4045061336
r10            0xefcdc810	4023240720
r11            0x3	3
r12            0x0	0
sp             0xf0ee3818	0xf0ee3818
lr             0x40	64
pc             0xf5d8ee78	0xf5d8ee78 <J9::ARM::AheadOfTimeCompile::initialiseAOTRelocationHeader(TR::IteratedExternalRelocation*)+1504>
cpsr           0x800f0010	-2146500592

```

**Comment by JamesKingdon** [comment]
Problem doesn't show up when JIT is built with debug, which makes things a little more challenging.

**Comment by knn-k** [comment]
The stmib instruction seems to be used in the following part of J9AheadOfTimeCompile.cpp:

```      case TR_ConstantPool:
      case TR_ConstantPoolOrderedPair:
      case TR_Trampolines:
      case TR_Thunks:
         {
         *(uintptrj_t *)cursor = (uintptrj_t)relocation->getTargetAddress2(); // inlined site index
         cursor += SIZEPOINTER;

         *(uintptrj_t *)cursor = (uintptrj_t)relocation->getTargetAddress(); // constantPool
         cursor += SIZEPOINTER;
         break;
         }
```

**Comment by JamesKingdon** [comment]
Awesome, thanks! How did you find it?
Now we have to choose how to fix it. I was hoping that annotating the declaration of cursor with the 'packed' attribute would work, but initial tests didn't seem to solve the problem. We could use memcpy or inline assembler for the moves as a slightly hacky alternative.

**Comment by knn-k** [comment]
The function has a big switch-case.  I found the jump table of the switch had four references to the address of the case block which contained the stmib instruction.  So I searched for blocks which had four case labels.

I don't think it is a good coding style, but an idea is to make the case block a fall-through to the next block as follows:
```
      case TR_ConstantPool:
      case TR_ConstantPoolOrderedPair:
      case TR_Trampolines:
      case TR_Thunks:
         {
         *(uintptrj_t *)cursor = (uintptrj_t)relocation->getTargetAddress2(); // inlined site index
         cursor += SIZEPOINTER;
         /* fall through */
         }
      case TR_CheckMethodEnter:
      case TR_CheckMethodExit:
         {
         *(uintptrj_t*)cursor = (uintptrj_t)relocation->getTargetAddress();
         cursor += SIZEPOINTER;
         }
        break;
```

**Comment by knn-k** [comment]
Yes, I agree.  My idea yesterday is not good for maintenance in the future.
I searched for GCC option that avoids using stm instructions (and other multi-word mem ops) in aarch32 code generation when 16-byte alignment is not guaranteed, but I could not find one.

**Comment by JamesKingdon** [comment]
@knn-k 
This is what I'm currently testing, what do you think?
```
         uintptrj_t value1 = (uintptrj_t)relocation->getTargetAddress2(); // inlined site index
         uintptrj_t value2 = (uintptrj_t)relocation->getTargetAddress(); // constantPool

         asm volatile( 
            "mov %[cursor], %[value1]\n\t"
            "add %[cursor], %[cursor], #4\n\t"
            "mov %[cursor], %[value2]\n\t"
            "add %[cursor], %[cursor], #4\n\t"
             :[cursor] "=r" (cursor)
             :[value1] "r" (&value1),[value2] "r" (&value2)
         );
```
It seems to work so far (it fixes at least one of the tests that was crashing before, and I've launched another run of the openj9 test suite). Obviously inline assembler is not ideal for maintainability, but I think the above is fairly clear? And we can add a comment block to explain the reasons for using it of course.

I suspect that we have to do the same thing for every block in that switch statement that does consecutive writes to *cursor as I don't see why the problem would affect one and not others.

**Comment by JamesKingdon** [comment]
Hmm, worryingly, when I try and use the same technique for the interpreter profiling data I get crashes, which suggests I still don't understand the inline assembler syntax and the above may be wrong.

**Comment by knn-k** [comment]
I don't have much experiences with inline assembler, and I cannot point out what to do with it.

I wondered why `cursor` had to point to unaligned address, but it does not look trivial to change the relocation data to aligned.

**Comment by knn-k** [comment]
I don't have much experiences with inline assembler, and I cannot point out what to do with it.

I wondered why `cursor` had to point to unaligned address, but it does not look easy to change the relocation data to aligned.

**Comment by JamesKingdon** [comment]
Yes, these unaligned buffers in interpreter profiling and AOT are frustrating, but I assume they are considered necessary for footprint density.
I wasn't confident in my assembler so I've been testing with a conservative change that uses memcpy instead: https://github.com/JamesKingdon/openj9/commit/944063d26893f15c16f1583860ee8124a478c267

**Comment by DanHeidinga** [comment]
Isn't there a compiler option to handle this?  GCC has `-mno-unaligned-access` and the usual mtune / mcpu options which enable or disable that by default.

https://gcc.gnu.org/onlinedocs/gcc/ARM-Options.html

**Comment by andrewcraik** [comment]
I'm not sure that the option is enough to take care of all the allocators in the JIT since we have our own bump pointer allocators for anything non-persistent it is possible that things could squeak past though I agree an option like `-mno-unaligned-access` would make things easier.

**Comment by JamesKingdon** [comment]
Sadly, I don't think the options apply. It's not that ARM can't do unaligned access, it can and we need to do so if we're not going to lose even more performance. It's just that there are some specific instructions that don't support unaligned access that gcc is generating inappropriately, possibly because of the very old gcc we're using, or possibly because we cast our unaligned pointers so heavily that gcc can no longer tell that they are unaligned.

**Comment by JamesKingdon** [comment]
I was hoping that I could use __attribute__ ((packed)) to mark specific pointers as unaligned but I wasn't able to make it work.

**Comment by andrewcraik** [comment]
I did have to help with some alignment problems on SPARC for LLVM at one point and for the bump pointer allocators the allocator had to be modified to align the allocation. It increased footprint, but it was the only way for sanity to prevail in the end. Not sure if that is the right call here, but something worthy of consideration.

**Comment by JamesKingdon** [comment]
Who is the architect for the AOTRelocationHeader - @mstoodle ?

**Comment by JamesKingdon** [comment]
(and we'd really need to have the same conversation about the interpreter profiling data)

**Comment by mstoodle** [comment]
@JamesKingdon Yes, I suppose that's me, though I'm not sure I want to call that code "architected" in any really positive sense of the word, although there are carefully designed elements. The relocation *runtime*, at least, has been architected in such a way that you'll *probably* be able to deal with it relatively cleanly via the RelocationTarget. 

Unfortunately, the compile side of this stuff never went through the re-engineering effort we did on the runtime side, and so we're left with lots of *(uintptrj_t *)cursor = stuff as opposed to the (cleaner, IMO, since I created it) the RelocationTarget support in the runtime.

The goal for that code is certainly footprint density for the relocation data.

I would not be in favour of @knn-k 's modification to split the two across different cases to prevent the merging of the writes, at least not as a longer term solution.

After chatting with James privately for a bit, I discovered that he already has a solution for this problem employing memcpy, so maybe there isn't more to discuss here atm?

**Comment by JamesKingdon** [comment]
I had a concern with the memcpy approach that future compilers would apply the one more step of optimisation apparently required to get back to generating store multiples again, so I've replaced all the problem areas with inline assembly. Fortunately this only impacts an ARM specific file, so any perceived ugliness is contained.
